lib/ReactTags.js

```diff
@@ -102,11 +106,32 @@ module.exports = React.createClass({
         // end up with a terminating character typed in.
         if (this.props.delimiters.indexOf(e.keyCode) !== -1) {
             if (e.keyCode !== Keys.TAB || query) {
-              e.preventDefault();
+                e.preventDefault();
             }
 
-            if (this.state.selectedIndex > -1) {
-                this.addTag(this.state.suggestions[this.state.selectedIndex]);
+            let newTag;
+            if (query) {
+                if (selectionMode) {
+                    newTag = suggestions[selectedIndex];
+                } else {
+                    // If there is a suggestion with the same name we use that one,
+                    // because the newly "typed in" one is missing all other properties.
+                    newTag = suggestions.find(suggestion =>
+                      suggestion.name === query  // tag names are unique
+                    );
+                    if (newTag === undefined) {
+                        if (this.props.allowNew) {
+                            // If new tags are allowed, add it with just the name.
+                            // The user is responsible for adding any other relevant
+                            // properties, like an Id.
+                            newTag = { name: query };
+                        } else {
+                            // Otherwise stop right here if no suitable suggestions is found.
+                            return;
+                        }
+                    }
+                }
+                this.addTag(newTag);
```

Comment:
I think this can be simplified to:

```js
if (this.state.selectedIndex > -1) {
  this.addTag(this.state.suggestions[this.state.selectedIndex]);
} else if (this.props.allowNew && query.length >= this.props.minQueryLength) {
  this.addTag(suggestions.find(suggestion => suggestion.name === query) || { name: query });
}
```